### PR TITLE
Added support for C# benchmarks

### PR DESCRIPTION
**Pull request by helehex**
Only allows C# benchmarks if the dotnet version of the engine is being used.
Tested on, and defaulted to vanilla Godot (without the mono module).
Tested through CLI.

Added:
- C# benchmark support
- `TestID.language` property
- `Manager.languages` dictionary
- Control (noop) benchmarks for both GDScript and C#
- Some C# benchmarks: ForLoop, Array, List, RigidBody3D
- Note about C# in the `CONTRIBUTING.md` doc
- Additional pruned CLI output after each test
- Case insensitive CLI benchmark masking

Notes:
- The base class of the method being timed is no longer strongly typed at the call site
- I used PascalCase for C# Benchmark functions.
- I kept snake_case for the test_ properties in the base Benchmark class

### CODE REVIEW

**Comment by franciscod** [comment]
What happens if two tests have the same name, something like `performance.gd` and `Performance.cs`? Would both run and be listed?

When running without .NET support, are the tests silently ignored or explicitely skipped? (what's in the JSON?)

**Comment by helehex** [comment]
I did a few more things:
- Wrote a note on the CONTRIBUTING.md document about C# and made a few spots more clear.
- Removed the CS suffix from .cs files. Not sure if I like this... but it doesn't really matter and still works fine.
- Added similar benchmarks to the array.gd tests, but using C# arrays and lists instead.

File: manager.gd [hunk 1]
```diff
@@ -122,6 +120,7 @@ func benchmark(test_ids: Array[TestID], return_path: String) -> void:
 	if return_path:
 		get_tree().change_scene_to_file(return_path)
 	else:
+		get_tree().queue_delete(get_tree())
```
Comment (helehex) [inline manager.gd:123]: This fixes a warning on exit: `ObjectDB Instances Leaked At Exit`.
I'm not sure if this is the proper way to fix it though.

**Comment by helehex** [comment]
I changed the CLI `--include-benchmarks` and `--exclude-benchmarks` args to be case insensitive at the comparison site. This alleviates the case differences in .cs files (has no effect when running through GUI). I'm not sure about this change, but it's easy to go back if necessary.

**Comment by helehex** [comment]
there's a problem with letting the bodies settle down during the benchmark:
because of the 5 second benchmark limit, there is no guarantee the same amount of physics frames will run so I'm getting less consistent results than I would like.

Also this should probably be a separate PR at this point, but since I'm still comparing C# and GDScript, I guess it makes sense.

**Comment by helehex** [comment]
it's a little more consistent, but still could be improved. The spheres are smaller and more rolley than boxes which causes them to scale with numbers/density differently than boxes. I guess just tune the constant to whatever feels right.

**Comment by helehex** [comment]
My RigidBody2D's are acting more like soup

**Comment by helehex** [comment]
soup bias

**Comment by helehex** [comment]
I'll leave the default physics2d settings event though it's a little soupy.
I'm still concerned with the inconsistency of the physics performance measurements, maybe due to the readings being unaligned with the start of the benchmark?

**Comment by Calinou** [comment]
> maybe due to the readings being unaligned with the start of the benchmark?

Indeed, this is a known issue: https://github.com/godotengine/godot-proposals/issues/6809

File: manager.gd (hunk 1 above)
Comment (helehex) [reply manager.gd:128]: only fixes it in certain situations, so maybe just remove that line

**Comment by helehex** [comment]
aside from language features benchmarks, it may be better to have a few general benchmarks like calling C# godot functions, instead of specific things like C# Rigidbody, to avoid having to many unnecessary things.

**Comment by helehex** [comment]
I merged upstream changes and resolved the conflict of benchmark.gd type check (and not benchmark.cs) by reverting the change and moving supplemental assets to a separate folder instead

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. When using a non-C# build, the project still runs correctly but C# benchmarks won't be listed or run.

I'll need to update the build scripts on the server so it compiles a .NET-enabled build.

**Pull request merged**